src/git.rs

```diff
@@ -1,263 +1,250 @@
+                        let file_path = file_path.to_string_lossy().to_string();
+
+                        if file_path.ends_with(".rs") {
+                            sections.push(Section {
+                                file_name: file_path,
+                                line_start: hunk.new_start(),
+                                line_end: hunk.new_start() + hunk.new_lines(),
+                            });
+                        }
                     }
                 }
-                // Resets start and end for next section
-                line_start = 0;
-                line_end = 0;
+                _ => {}
             }
-        }
-        if line_start != 0 && file_name != "" {
-            if let Some(s) = create_section(&file_name, line_start, line_end) {
-                sections.push(s);
-            }
-        }
-        sections
-    }
-}
+            true
+        }),
+        None,
+    )?;
 
-fn create_section(file_name: &str, line_start: i32, line_end: i32) -> Option<Section> {
-    let mut current_section = SectionBuilder::new();
-    current_section.file_name(file_name.to_string());
-    current_section.line_start(line_start);
-    current_section.line_end(line_end);
-    current_section.build()
-}
-
-fn get_diff_line_start(line: &str) -> i32 {
-    // @@ and space
-    let after_ats = &line[3..];
-    // space and @@
-    let before_second_ats_index = &after_ats.find("@@").unwrap() - 1;
-    // -33,6 +33,9
-    let diff_lines = &after_ats[..before_second_ats_index];
-    let (_, to) = diff_lines.split_at(diff_lines.find(' ').unwrap());
-    let added = to.trim();
-    let (added_start, _) = if let Some(index) = added[1..].find(',') {
-        let (a, b) = added[1..].split_at(index);
-        (a, &b[1..])
-    } else {
-        (added, "")
-    };
-    added_start.parse::<i32>().unwrap()
+    Ok(sections)
 }
 
 #[cfg(test)]
 mod tests {
-    #[test]
-    fn test_set_verbose() {
-        use crate::git::Parser;
+    use super::*;
+    use std::fs::{self, File};
+    use std::io::Write;
+    use tempfile::TempDir;
 
-        let mut parser = Parser::new();
-        assert_eq!(false, parser.verbose);
+    type Result<T> = std::result::Result<T, Error>;
 
-        let p2 = parser.set_verbose(true);
-        assert_eq!(true, p2.verbose);
-
-        let p3 = p2.set_verbose(false);
-        assert_eq!(false, p3.verbose);
-    }
     #[test]
-    fn test_empty_diff() {
-        use crate::git::{Parser, Section};
-        // Setup
-        let diff = r#""#;
-        let expected_sections: Vec<Section> = vec![];
-        let parser = Parser::new();
-        // Run
-        let actual_sections = parser.sections(diff);
-        // Assert
-        assert_eq!(expected_sections, actual_sections);
+    fn no_changes() -> Result<()> {
+        let repo = RepoFixture::new()?;
+
+        let sections = get_sections(repo.path(), "master").unwrap();
+        assert!(sections.is_empty());
+        Ok(())
```

Comment:
That's interesting, I have never thought of returning results in tests.
What would be the advantage?

Reply:
Returning a `Result` allows us to use the question mark operator instead of explicitly calling `unwrap`. I think it makes the code in the tests more readable since there is a lot of I/O that could fail.

Reply:
It's much more readable indeed! Let's use it in cargo-scout from now on 🎉